ArduCopter/mode.cpp

```diff
@@ -431,6 +431,33 @@ void Mode::get_pilot_desired_lean_angles(float &roll_out, float &pitch_out, floa
     // roll_out and pitch_out are returned
 }
 
+// get_pilot_vel_correction - transform pilot's roll or pitch input into a desired velocity
+Vector2f Mode::get_pilot_vel_correction(float vel_max) const
+{
+    Vector2f vel;
+
+    // throttle failsafe check
+    if (copter.failsafe.radio || !copter.ap.rc_receiver_present) {
+        return vel;
+    }
+    // fetch roll and pitch inputs
+    float roll_out = channel_roll->get_control_in();
+    float pitch_out = channel_pitch->get_control_in();
+
+    // scale roll and pitch inputs to ANGLE_MAX parameter range
+    float scaler = 1.0 / (float)ROLL_PITCH_YAW_INPUT_MAX;
+    roll_out *= scaler;
+    pitch_out *= scaler;
+
+    vel = Vector2f(pitch_out, roll_out);
+    if (vel.is_zero()) {
+        return vel;
+    }
+    Vector2f vel_scaler = vel / MAX(fabsf(vel.x), fabsf(vel.y));
+    vel *= vel_max / vel_scaler.length();
+    return vel;
```

Comment:
Why not this?
```suggestion
    return vel.normalized() * vel_max;
```

Reply:
I have added comments to explain:

    // Transform square input range to circular output
    // vel_scaler is the vector to the edge of the +- 1.0 square in the direction of the current input
    Vector2f vel_scaler = vel / MAX(fabsf(vel.x), fabsf(vel.y));
    // We scale the output by the ratio of the distance to the square to the unit circle and multiply by vel_max
    vel *= vel_max / vel_scaler.length();